prow/hook/server.go

```diff
@@ -150,3 +166,58 @@ func (s *Server) demuxEvent(eventType, eventGUID string, payload []byte) error {
 	}
 	return nil
 }
+
+// demuxExternal dispatches events to external plugins. The headers originating
+// in the Github webhook are copied as is in the new requests.
+func (s *Server) demuxExternal(eventType, eventGUID string, payload []byte, h http.Header) {
+	l := logrus.WithFields(logrus.Fields{"event-type": eventType, "event-GUID": eventGUID})
+	for _, p := range s.filterExternalPlugins(eventType) {
+		go func(p plugins.ExternalPlugin) {
+			if err := s.dispatch(p.Service, payload, h); err != nil {
+				l.WithError(err).Errorf("Error dispatching event to external plugin %q.", p.Name)
+			}
+		}(p)
+	}
+}
+
+// Filter external plugins by event type.
+func (s *Server) filterExternalPlugins(eventType string) []plugins.ExternalPlugin {
+	var affected []plugins.ExternalPlugin
+	for _, ext := range s.Plugins.Config().ExternalPlugins {
+		for _, event := range ext.Events {
+			if event == eventType {
+				affected = append(affected, ext)
+				break
+			}
+		}
+	}
+	return affected
+}
+
+// dispatch creates a new request using the provided payload and headers
+// and dispatches the request to a Kubernetes service.
+func (s *Server) dispatch(hostname string, payload []byte, h http.Header) error {
+	// The following URL should translate to a DNS name resolvable inside
+	// the cluster.
+	url := fmt.Sprintf("http://%s", hostname)
+	buf := bytes.NewBuffer(payload)
+	req, err := http.NewRequest(http.MethodPost, url, buf)
+	if err != nil {
+		return err
+	}
+	req.Header = h
+	req.Header.Set("User-Agent", "ProwHook")
+	resp, err := s.c.Do(req)
+	if err != nil {
```

Comment:
Consider retries?

Reply:
Done. Not sure about retrying on 500s though